core/evm.go

```diff
@@ -501,3 +515,137 @@ func Transfer(db vm.StateDB, sender, recipient common.Address, amount *big.Int,
 		db.AddBalance(recipient, amount)
 	}
 }
+
+// FetchStakingInfoFn is the EVM implementation of the read only staking precompile
+// avoid import cycle by putting this function here
+func FetchStakingInfoFn(ref *block.Header, chain ChainContext) vm.RoStakingPrecompileFunc {
+	return func(db vm.StateDB, roStakeMsg *stakingTypes.ReadOnlyStakeMsg) ([]byte, error) {
+		switch roStakeMsg.What {
+		case "DelegationByDelegatorAndValidator":
+			{
+				// either we iterate over ReadDelegationsByDelegator
+				// or over wrapper.Delegations. since ReadDelegationsByDelegator
+				// is only written at the end of a block, prefer wrapper.Delegations
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				for _, delegation := range wrapper.Delegations {
+					if bytes.Equal(
+						delegation.DelegatorAddress[:],
+						roStakeMsg.DelegatorAddress[:],
+					) {
+						return common.LeftPadBytes(
+							delegation.Amount.Bytes(), 32,
+						), nil
+					}
+				}
+				return big.NewInt(0).Bytes(), nil
+			}
+		case "ValidatorMaxTotalDelegation":
+			{
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				return common.LeftPadBytes(
+					wrapper.MaxTotalDelegation.Bytes(), 32,
+				), nil
+			}
+		case "ValidatorTotalDelegation":
+			{
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				return common.LeftPadBytes(
+					wrapper.TotalDelegation().Bytes(), 32,
+				), nil
+			}
+		case "BalanceAvailableForRedelegation":
```

Comment:
Why is this data needed?

Reply:
Are we still sure this data is needed? We want to keep the onchain logic as simple as possible. Anything that can be done offchain should not go onchain.